Autoscaler killing busy runners again — #640 symptom recurred, blocking UpstreamDrift PR backlog

## Summary
The symptom from #640 (*Autoscaler kills busy runners mid-job*) has **recurred** as of 2026-05-29 UTC, despite #640 being closed via #664 (job-pickup race / orphan-Worker gap) and #679 (graceful-drain drop-in, "Fix 3"). Jobs are being terminated mid-run with `The runner has received a shutdown signal ... The operation was canceled`, and the fleet shows **offline-but-still-`busy`** orphaned runners.

## Evidence (2026-05-29 UTC)
- Org runner snapshot fluctuating between **1–6 `offline` + `busy=true`** orphaned runners with only **1 idle** runner; the rest `online`/`busy`. Observed orphans included: `d-sorg-local-Desktop-5`, `d-sorg-local-Oglaptop-7` (and earlier 6 others).
- Failing jobs show, during `actions/checkout`:
  ```
  ##[error]The runner has received a shutdown signal. This can happen when the runner service is stopped, or a manually started runner is canceled.
  ##[error]The operation was canceled.
  ```
- Runner state-corruption signature also present: `git fsck failed, purging .git...` at job start (matches the known corruption pattern).

## Impact
Blocks the entire open-PR backlog in **UpstreamDrift** (13 PRs). The only ruleset-required check, `quality-gate`, cannot stay green because runners are killed mid-job. Concrete example: **UpstreamDrift#6621** (a trivial, locally-validated CI fix) had its `quality-gate` killed during checkout (~90s in).

## Likely cause
Either (a) a regression of #640, or (b) the #664/#679 graceful-drain drop-in is **merged but not deployed** to the live fleet at `~/actions-runners/dashboard/` (non-root deploy required the autoscaler lock fix from #664).

## Requested actions
1. Verify the #679 graceful-drain systemd drop-in for `actions.runner.*` units is actually **installed and active** on the live runner hosts (not just merged).
2. Drain / de-register the orphaned `offline+busy` runners so jobs stop routing to dead workers.
3. Confirm the autoscaler is not scaling down runners that are mid-job (re-check `_runner_is_busy()` / `MainPID=0` transient handling from #640).

Filed during a paused PR-remediation run; the UpstreamDrift PR drive will resume once the fleet is stable.

--- a/.github/workflows/anti-phantom-merge.yml
+++ b/.github/workflows/anti-phantom-merge.yml
@@ -47,10 +47,7 @@ jobs:
             fi
           fi
 
-          STATS=$(gh pr view "$PR" --repo "$REPO" --json additions,deletions,changedFiles)
-          ADDS=$(echo "$STATS" | jq -r .additions)
-          DELS=$(echo "$STATS" | jq -r .deletions)
-          FILES=$(echo "$STATS" | jq -r .changedFiles)
+          FILES=$(gh pr view "$PR" --repo "$REPO" --json changedFiles --jq .changedFiles)
           TITLE=$(gh pr view "$PR" --repo "$REPO" --json title --jq .title)
           BODY=$(gh pr view "$PR" --repo "$REPO" --json body --jq .body)
           PR_FILES=$(

--- a/SPEC.md
+++ b/SPEC.md
@@ -1,9 +1,18 @@
 # SPEC.md â€” D-sorganization Runner Dashboard
 
-**Spec Version:** 2.5.42
+**Spec Version:** 2.5.43
 **Application Version:** 4.1.3 (see `VERSION`)
 **Last Updated:** 2026-05-29T00:00:00-07:00
 **Status:** Active
+
+- **2026-05-29 (2.5.43):** Hardened autoscaler scale-down against undeployed
+  runner drain drop-ins (issue #785). Before `backend/autoscaler_systemd.py`
+  calls `systemctl stop` for any `actions.runner.*` unit, it now verifies the
+  effective systemd stop contract reports `KillMode=mixed` and
+  `TimeoutStopUSec >= 120s`. If a host has the code but not the active
+  #640/#679 drop-in, the autoscaler refuses to stop the runner and logs the
+  deploy action instead of risking a mid-job shutdown. Focused tests cover the
+  systemd time-span parser, safe/unsafe contract decisions, and the stop guard.
 
 - **2026-05-28 (2.5.40):** Storage-tier aware disk pressure metrics and classification (issue #754).
   `backend/system_utils.py` gains two new pure helpers:

--- a/backend/autoscaler_systemd.py
+++ b/backend/autoscaler_systemd.py
@@ -8,6 +8,7 @@ shell=True) and honour the configured timeout from autoscaler_config.
 from __future__ import annotations
 
 import logging
+import re
 import subprocess
 import sys
 
@@ -17,6 +18,7 @@ from autoscaler_config import (
 )
 
 log = logging.getLogger("runner-autoscaler")
+_MIN_SAFE_STOP_TIMEOUT_SECONDS = 120
 
 
 def _dry_run_enabled() -> bool:
@@ -109,6 +111,72 @@ def _runner_workdir_for_unit(unit: str) -> str:
     return (r.stdout or "").strip()
 
 
+def _systemd_timespan_seconds(value: str) -> float:
+    """Parse common systemd time spans into seconds."""
+    text = value.strip()
+    if not text:
+        return 0.0
+    if text.isdigit():
+        raw_value = float(text)
+        return raw_value / 1_000_000 if raw_value > 10_000 else raw_value
+
+    total = 0.0
+    multipliers = {
+        "usec": 0.000001,
+        "us": 0.000001,
+        "ms": 0.001,
+        "s": 1.0,
+        "min": 60.0,
+        "h": 3600.0,
+        "d": 86400.0,
+    }
+    for amount, unit in re.findall(r"(\d+(?:\.\d+)?)\s*(usec|us|ms|s|min|h|d)", text):
+        total += float(amount) * multipliers[unit]
+    return total
+
+
+def _unit_has_safe_stop_contract(unit: str) -> bool:
+    """Return True when stopping *unit* should not kill active jobs.
+
+    The autoscaler may only stop runner units whose effective systemd
+    configuration has the #640/#679 drain contract loaded. If the host was not
+    redeployed and still has the unsafe default, refusing scale-down is safer
+    than terminating a checkout or test step mid-job.
+    """
+    try:
+        r = subprocess.run(
+            ["systemctl", "show", unit, "--property=KillMode,TimeoutStopUSec"],
+            capture_output=True,
+            text=True,
+            timeout=_SYSTEMCTL_TIMEOUT_S,
+            check=False,
+        )
+    except (OSError, subprocess.TimeoutExpired) as exc:
+        log.warning("Could not inspect stop contract for %s: %s", unit, exc)
+        return False
+
+    props: dict[str, str] = {}
+    for line in (r.stdout or "").splitlines():
+        if "=" in line:
+            key, value = line.split("=", 1)
+            props[key] = value.strip()
+
+    kill_mode = props.get("KillMode", "")
+    timeout_seconds = _systemd_timespan_seconds(props.get("TimeoutStopUSec", ""))
+    if kill_mode != "mixed" or timeout_seconds < _MIN_SAFE_STOP_TIMEOUT_SECONDS:
+        log.error(
+            "Refusing to stop %s: unsafe runner stop contract "
+            "(KillMode=%s TimeoutStopUSec=%s, need KillMode=mixed and timeout>=%ss). "
+            "Run deploy/install-runner-maintenance.sh or deploy/migrate-runner-units.sh.",
+            unit,
+            kill_mode or "?",
+            props.get("TimeoutStopUSec", "?"),
+            _MIN_SAFE_STOP_TIMEOUT_SECONDS,
+        )
+        return False
+    return True
+
+
 def _stop_unit(unit: str) -> bool:
     """Stop *unit* via systemd, respecting DRY_RUN mode.
 
@@ -122,6 +190,8 @@ def _stop_unit(unit: str) -> bool:
     if _dry_run_enabled():
         log.info("[dry-run] would stop %s", unit)
         return True
+    if not _unit_has_safe_stop_contract(unit):
+        return False
     r = subprocess.run(
         ["sudo", "-n", "systemctl", "stop", unit],
         capture_output=True,

--- a/backend/runner_autoscaler.py
+++ b/backend/runner_autoscaler.py
@@ -120,6 +120,7 @@ from autoscaler_systemd import (
     _runner_workdir_for_unit,
     _start_unit,
     _stop_unit,
+    _unit_has_safe_stop_contract,
     _unit_is_active,
     _unit_state,
 )
@@ -161,6 +162,7 @@ __all__ = [
     "_runner_workdir_for_unit",
     "_start_unit",
     "_stop_unit",
+    "_unit_has_safe_stop_contract",
     "_unit_is_active",
     "_unit_state",
     "_runner_busy_via_lockfile",

--- a/tests/test_autoscaler_systemd.py
+++ b/tests/test_autoscaler_systemd.py
@@ -77,6 +77,21 @@ class TestUnitMetadata:
         with patch("subprocess.run", return_value=_cp("/srv/actions/runner\n")):
             assert sd._runner_workdir_for_unit("some.service") == "/srv/actions/runner"
 
+    def test_safe_stop_contract_accepts_mixed_with_long_timeout(self) -> None:
+        stdout = "KillMode=mixed\nTimeoutStopUSec=2min\n"
+        with patch("subprocess.run", return_value=_cp(stdout)):
+            assert sd._unit_has_safe_stop_contract("actions.runner.org.r.service") is True
+
+    def test_safe_stop_contract_rejects_process_kill_mode(self) -> None:
+        stdout = "KillMode=process\nTimeoutStopUSec=1min 30s\n"
+        with patch("subprocess.run", return_value=_cp(stdout)):
+            assert sd._unit_has_safe_stop_contract("actions.runner.org.r.service") is False
+
+    def test_safe_stop_contract_rejects_short_timeout(self) -> None:
+        stdout = "KillMode=mixed\nTimeoutStopUSec=90s\n"
+        with patch("subprocess.run", return_value=_cp(stdout)):
+            assert sd._unit_has_safe_stop_contract("actions.runner.org.r.service") is False
+
 
 class TestDryRunEnabled:
     def test_module_flag_enables_dry_run(self, monkeypatch: pytest.MonkeyPatch) -> None:
@@ -93,7 +108,13 @@ class TestDryRunEnabled:
 class TestStopUnit:
     def test_happy_invokes_cleanup(self) -> None:
         with (
-            patch("subprocess.run", return_value=_cp(returncode=0)),
+            patch(
+                "subprocess.run",
+                side_effect=[
+                    _cp("KillMode=mixed\nTimeoutStopUSec=2min\n"),
+                    _cp(returncode=0),
+                ],
+            ),
             patch("runner_state_cleanup.cleanup_runner_state") as mock_clean,
         ):
             assert sd._stop_unit("actions.runner.org.r.service") is True
@@ -108,7 +129,13 @@ class TestStopUnit:
         the runner had already died. Cleanup must run on every stop path.
         """
         with (
-            patch("subprocess.run", return_value=_cp(returncode=1)),
+            patch(
+                "subprocess.run",
+                side_effect=[
+                    _cp("KillMode=mixed\nTimeoutStopUSec=2min\n"),
+                    _cp(returncode=1),
+                ],
+            ),
             patch("runner_state_cleanup.cleanup_runner_state") as mock_clean,
         ):
             assert sd._stop_unit("actions.runner.org.r.service") is False
@@ -117,13 +144,28 @@ class TestStopUnit:
     def test_cleanup_error_does_not_break_stop(self) -> None:
         """If cleanup itself raises, _stop_unit still returns the stop status."""
         with (
-            patch("subprocess.run", return_value=_cp(returncode=0)),
+            patch(
+                "subprocess.run",
+                side_effect=[
+                    _cp("KillMode=mixed\nTimeoutStopUSec=2min\n"),
+                    _cp(returncode=0),
+                ],
+            ),
             patch(
                 "runner_state_cleanup.cleanup_runner_state",
                 side_effect=RuntimeError("cleanup boom"),
             ),
         ):
             assert sd._stop_unit("actions.runner.org.r.service") is True
+
+    def test_refuses_stop_when_unit_lacks_safe_stop_contract(self) -> None:
+        with (
+            patch("subprocess.run", return_value=_cp("KillMode=process\nTimeoutStopUSec=90s\n")) as mock_run,
+            patch("runner_state_cleanup.cleanup_runner_state") as mock_clean,
+        ):
+            assert sd._stop_unit("actions.runner.org.r.service") is False
+            assert mock_run.call_count == 1
+            mock_clean.assert_not_called()
 
     def test_dry_run(self, monkeypatch: pytest.MonkeyPatch) -> None:  # type: ignore[name-defined]
         monkeypatch.setattr(sd, "DRY_RUN", True)

--- a/tests/test_runner_autoscaler.py
+++ b/tests/test_runner_autoscaler.py
@@ -116,7 +116,13 @@ def test_list_runner_units_timeout() -> None:
 
 
 def test_stop_unit_happy() -> None:
-    with patch("subprocess.run", return_value=_make_completed_process("", returncode=0)):
+    with patch(
+        "subprocess.run",
+        side_effect=[
+            _make_completed_process("KillMode=mixed\nTimeoutStopUSec=2min\n", returncode=0),
+            _make_completed_process("", returncode=0),
+        ],
+    ):
         result = ra._stop_unit("actions.runner.my-org.runner1.service")
     assert result is True
 
@@ -124,7 +130,13 @@ def test_stop_unit_happy() -> None:
 def test_stop_unit_sudo_failure() -> None:
     """The sudo-failure path: returncode != 0 returns False, does not raise."""
     cp = _make_completed_process("sudo: permission denied", returncode=1)
-    with patch("subprocess.run", return_value=cp):
+    with patch(
+        "subprocess.run",
+        side_effect=[
+            _make_completed_process("KillMode=mixed\nTimeoutStopUSec=2min\n", returncode=0),
+            cp,
+        ],
+    ):
         result = ra._stop_unit("actions.runner.my-org.runner1.service")
     assert result is False
 

